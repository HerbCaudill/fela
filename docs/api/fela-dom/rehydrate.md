# `rehydrate(renderer)`

Rehydrates the renderer's cache from existing `<style>` elements.<br>It is primarily used to rehydrate the cache from server-rendered CSS.

## Arguments
1. `renderer` ([*Renderer*](../fela/Renderer.md)): The renderer which's cache is rehydrated.


## Example

Assuming we have the following server-rendered elements:
```HTML
<style data-fela-type="RULE">.a{color:red}.b{color:blue}</style>
```

```javascript
import { createRenderer } from 'fela'
import { rehydrate } from 'fela-dom'

const renderer = createRenderer()

rehydrate(renderer)

const rule = props => ({
  color: props.color
})

renderer.renderRule(rule, { color: 'red' }) // => a
renderer.renderRule(rule, { color: 'green' }) // => c
renderer.renderRule(rule, { color: 'blue' }) // => b
```
