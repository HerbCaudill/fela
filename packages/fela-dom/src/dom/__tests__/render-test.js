import { html as beautify } from 'js-beautify'
import { createRenderer } from 'fela'

import render from '../render'
import rehydrate from '../rehydration/rehydrate'
import renderToMarkup from '../../server/renderToMarkup'

beforeEach(() => {
  const head = document.head
  while (head.firstChild) {
    head.removeChild(head.firstChild)
  }
})

describe('render', () => {
  it('should create style nodes and render CSS rules', () => {
    const renderer = createRenderer()
    renderer.renderRule(() => ({
      backgroundColor: 'red',
      color: 'blue',
    }))
    renderer.renderKeyframe(() => ({
      '0%': {
        color: 'yellow',
      },
      '100%': {
        color: 'orange',
      },
    }))
    renderer.renderFont('Lato', ['../Lato.ttf'], {
      fontWeight: 300,
    })
    render(renderer)
    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should correctly sort rules', () => {
    process.env.NODE_ENV = 'production'

    const renderer = createRenderer()
    render(renderer)

    renderer.renderRule(() => ({
      color: 'blue',
      ':focus-within': {
        color: 'black',
      },
      ':hover': {
        color: 'red',
      },
      ':active': {
        color: 'yellow',
      },
    }))

    renderer.renderRule(() => ({
      color: 'red',
      ':hover': {
        color: 'blue',
      },
    }))

    process.env.NODE_ENV = 'development'

    expect(document.styleSheets[0].cssRules).toMatchSnapshot()
  })

  it('should not render multiple times', () => {
    const renderer = createRenderer()

    render(renderer)
    render(renderer)

    renderer.renderRule(() => ({
      backgroundColor: 'red',
      color: 'blue',
    }))

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })

  it('should not overwrite rehydrated styles', () => {
    const serverRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
    })

    serverRenderer.renderRule(() => ({
      color: 'yellow',
      ':hover': {
        color: 'red',
      },
      '@media (max-width: 800px)': {
        color: 'blue',
      },
    }))

    document.head.innerHTML = renderToMarkup(serverRenderer)

    const clientRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
    })

    rehydrate(clientRenderer)

    clientRenderer.renderRule(() => ({
      backgroundColor: 'red',
      ':hover': {
        color: 'red',
      },
      color: 'blue',
    }))

    render(clientRenderer)

    expect(
      beautify(document.documentElement.outerHTML, {
        indent_size: 2,
      })
    ).toMatchSnapshot()
  })
})
