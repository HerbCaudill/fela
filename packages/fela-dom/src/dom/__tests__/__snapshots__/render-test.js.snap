// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should correctly sort rules 1`] = `
Array [
  CSSStyleRule {
    "__ends": 14,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 0,
    "selectorText": ".a",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 2,
      "_importants": Object {
        "color": "",
      },
      "color": "blue",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 13,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 0,
    "selectorText": ".e",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 2,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 19,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 3,
    "selectorText": ".c:hover",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 8,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 20,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 3,
    "selectorText": ".f:hover",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 8,
      "_importants": Object {
        "color": "",
      },
      "color": "blue",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 28,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 4,
    "selectorText": ".b:focus-within",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 15,
      "_importants": Object {
        "color": "",
      },
      "color": "black",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 23,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "score": 6,
    "selectorText": ".d:active",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 9,
      "_importants": Object {
        "color": "",
      },
      "color": "yellow",
      "length": 1,
      "parentRule": [Circular],
    },
  },
]
`;

exports[`render should create style nodes and render CSS rules 1`] = `
"<html>

<head>
  <style data-fela-type=\\"FONT\\" type=\\"text/css\\">
    @font-face {
      font-weight: 300;
      src: url('../Lato.ttf') format('truetype');
      font-family: \\"Lato\\"
    }
  </style>
  <style data-fela-type=\\"KEYFRAME\\" type=\\"text/css\\">
    @-webkit-keyframes k1 {
      0% {
        color: yellow
      }
      100% {
        color: orange
      }
    }

    @-moz-keyframes k1 {
      0% {
        color: yellow
      }
      100% {
        color: orange
      }
    }

    @keyframes k1 {
      0% {
        color: yellow
      }
      100% {
        color: orange
      }
    }
  </style>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\">
    .a {
      background-color: red
    }

    .b {
      color: blue
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`render should not overwrite rehydrated styles 1`] = `
"<html>

<head>
  <style type=\\"text/css\\" data-fela-rehydration=\\"4\\" data-fela-type=\\"RULE\\">
    .b {
      color: yellow
    }

    .e {
      background-color: red
    }

    .f {
      color: blue
    }

    .c:hover {
      color: red
    }
  </style>
  <style type=\\"text/css\\" data-fela-rehydration=\\"4\\" data-fela-type=\\"RULE\\" media=\\"(max-width: 800px)\\">
    .d {
      color: blue
    }
  </style>
</head>

<body></body>

</html>"
`;

exports[`render should not render multiple times 1`] = `
"<html>

<head>
  <style data-fela-type=\\"RULE\\" type=\\"text/css\\">
    .a {
      background-color: red
    }

    .b {
      color: blue
    }
  </style>
</head>

<body></body>

</html>"
`;
